x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/utils/XContentObjectTransformerTests.java

```diff
@@ -0,0 +1,85 @@
+
+public class XContentObjectTransformerTests extends ESTestCase {
+
+    public void testFromMap() throws IOException {
+        Map<String, Object> aggMap = Collections.singletonMap("fieldName",
+            Collections.singletonMap("max",
+                Collections.singletonMap("field", "fieldName")));
+
+        XContentObjectTransformer<AggregatorFactories.Builder> aggTransformer = XContentObjectTransformer.aggregatorTransformer();
+        assertXContentAreEqual(aggTransformer.fromMap(aggMap), aggMap);
+        assertXContentAreEqual(aggTransformer.fromMap(aggMap), aggTransformer.toMap(aggTransformer.fromMap(aggMap)));
+
+        Map<String, Object> queryMap = Collections.singletonMap("match",
+            Collections.singletonMap("fieldName", new HashMap<String, Object>(){{
+                // Add all the default fields so they are not added dynamically when the object is parsed
+                put("query","fieldValue");
+                put("operator","OR");
+                put("prefix_length",0);
+                put("max_expansions",50);
+                put("fuzzy_transpositions",true);
+                put("lenient",false);
+                put("zero_terms_query","NONE");
+                put("auto_generate_synonyms_phrase_query",true);
+                put("boost",1.0);
+            }}));
+
+        XContentObjectTransformer<QueryBuilder> queryBuilderTransformer = XContentObjectTransformer.queryBuilderTransformer();
+        assertXContentAreEqual(queryBuilderTransformer.fromMap(queryMap), queryMap);
+        assertXContentAreEqual(queryBuilderTransformer.fromMap(queryMap),
+            queryBuilderTransformer.toMap(queryBuilderTransformer.fromMap(queryMap)));
+    }
+
+    public void testToMap() throws IOException {
+        XContentObjectTransformer<AggregatorFactories.Builder> aggTransformer = XContentObjectTransformer.aggregatorTransformer();
+        XContentObjectTransformer<QueryBuilder> queryBuilderTransformer = XContentObjectTransformer.queryBuilderTransformer();
+
+        AggregatorFactories.Builder aggs = new AggregatorFactories.Builder();
+        long aggHistogramInterval = randomNonNegativeLong();
+        MaxAggregationBuilder maxTime = AggregationBuilders.max("time").field("time");
+        aggs.addAggregator(AggregationBuilders.dateHistogram("buckets")
+            .interval(aggHistogramInterval).subAggregation(maxTime).field("time"));
+
+        assertXContentAreEqual(aggs, aggTransformer.toMap(aggs));
+        assertXContentAreEqual(aggTransformer.fromMap(aggTransformer.toMap(aggs)), aggTransformer.toMap(aggs));
+
+        QueryBuilder queryBuilder = QueryBuilders.termQuery(randomAlphaOfLength(10), randomAlphaOfLength(10));
+
+        assertXContentAreEqual(queryBuilder, queryBuilderTransformer.toMap(queryBuilder));
+        assertXContentAreEqual(queryBuilderTransformer.fromMap(queryBuilderTransformer.toMap(queryBuilder)),
+            queryBuilderTransformer.toMap(queryBuilder));
+    }
+
```

Comment:
It would be good to also test a case of syntax that no longer works.  This is a query that was legal in 5.x but isn't now:

```
            "query" : {
               "match" : {
                 "airline" : {
                   "query" : "The type setting is deprecated in 5 and removed in 6",
                   "type" : "phrase"
                 }
               }
             }
```

So please add a test that asserts that that query can be successfully parsed to a map but then `fromMap` on that map throws an exception.